proposals/index-range-changes.md

```diff
@@ -0,0 +1,297 @@
+`receiver[expr]`.
+
+The first argument of `Slice` will be obtained by converting the typed expression in the following way:
+
+- When `expr` is of the form `expr1..expr2` (where `expr2` can be omitted) and `expr1` has type `int`, then it will be
+emitted as `expr1`.
+- When `expr` is of the form `^expr1..expr2` (where `expr2` can be omitted), then it will be emitted as 
+`receiver.Length - expr1`.
+- When `expr` is of the form `..expr2` (where `expr2` can be omitted), then it will be emitted as `0`.
+- Otherwise, it will be emitted as `expr.Start.GetOffset(receiver.Length)`.
+
+This value will be re-used in the calculation of the second `Slice` argument. When doing so it will be referred to as
+`start`. The second argument of `Slice` will be obtained by converting the range typed expression in the following way:
+
+- When `expr` is of the form `expr1..expr2` (where `expr1` can be omitted) and `expr2` has type `int`, then it will be 
+emitted as `expr2 - start`.
+- When `expr` is of the form `expr1..^expr2` (where `expr1` can be omitted), then it will be emitted as 
+`(receiver.Length - expr2) - start`.
+- When `expr` is of the form `expr1..` (where `expr1` can be omitted), then it will be emitted as `receiver.Length`.
+- Otherwise, it will be emitted as `expr.End.GetOffset(receiver.Length) - start`.
+
+The `receiver`, `Length` and `expr` expressions will be spilled as appropriate to ensure any side effects are only
+executed once. For example:
+
+``` csharp
+class Collection {
+    private int[] _array = new[] { 1, 2, 3 };
+
+    int Length {
+        get {
+            Console.Write("Length ");
+            return _array.Length;
+        }
+    }
+
+    int[] Slice(int start, int length) { 
+        var slice = new int[length];
+        Array.Copy(_array, start, slice, 0, length);
+        return slice;
+    }
+}
+
+class SideEffect {
+    Collection Get() {
+        Console.Write("Get ");
+        return new Collection();
+    }
+
+    void Use() { 
+        var array = Get()[0..2];
+        Console.WriteLine(array.length);
+    }
+}
+```
+
+This code will print "Get Length 2". 
+
+The language will special case the following known types: 
+
+- `string`: the method `Substring` will be used instead of `Slice`.
+- `array`: the method `System.Reflection.CompilerServices.GetSubArray` will be used instead of `Slice`.
+
+## Open Issues
+
+## Considerations
+
+### Detect Indexable based on ICollection
+The inspiration for this proposal was collection initializers. Using the structure of a type to convey that it had
+opted into a feature. In the case of collection initializers types can opt into the feature by implementing the 
+interface `IEnumerable` (non generic).
+
+This proposal initially required that types implement `ICollection` in order to qualify as Indexable. That required a number of special cases though:
+required a number of special cases:
```

Comment:
```suggestion
```
Guess I shouldn’t have written that one using a suggestion there that you can just apply to a single line 😅